clientv3/maintenance.go

```diff
@@ -145,6 +149,38 @@ func (m *maintenance) Status(ctx context.Context, endpoint string) (*StatusRespo
 	return (*StatusResponse)(resp), nil
 }
 
+func (m *maintenance) Snapshot(ctx context.Context) (io.ReadCloser, error) {
+	ss, err := m.getRemote().Snapshot(ctx, &pb.SnapshotRequest{})
+	if err != nil {
+		return nil, err
+	}
+
+	pr, pw := io.Pipe()
+	go func() {
+		for {
+			resp, err := ss.Recv()
+			if err != nil {
+				pw.CloseWithError(err)
+				return
+			}
+			if resp == nil && err == nil {
+				break
+			}
+			bw := 0
+			for bw < len(resp.Blob) {
+				n, werr := pw.Write(resp.Blob[bw:])
+				if werr != nil {
+					pw.CloseWithError(werr)
+					return
+				}
+				bw += n
+			}
+		}
+		pw.Close()
```

Comment:
There are failure paths that do `CloseWithError` so `defer pw.Close()` would try to double close.